Shared/Samples/Snap geometry edits with utility network rules/SnapGeometryEditsWithUtilityNetworkRulesView.swift

```diff
@@ -0,0 +1,300 @@
+private struct SnapRuleBehaviorLegend: View {
+    /// The images representing the different snap rule behavior cases.
+    let images: [SnapRuleBehavior?: Image]
+    
+    var body: some View {
+        LabeledContent("Snapping") {
+            HStack {
+                ForEach(SnapRuleBehavior?.allCases, id: \.self) { behavior in
+                    Label {
+                        Text(behavior.label)
+                    } icon: {
+                        images[behavior]
+                    }
+                }
+            }
+        }
+        .font(.footnote)
+    }
+}
+
+/// A list for enabling and disabling snap source settings.
+private struct SnapSourcesList: View {
+    /// The snap source settings to show in the list.
+    let settings: [SnapSourceSettings]
+    
+    /// The images representing the different snap rule behavior cases.
+    let ruleBehaviorImages: [SnapRuleBehavior?: Image]
+    
+    /// The action to dismiss the view.
+    @Environment(\.dismiss) private var dismiss
+    
+    var body: some View {
+        NavigationStack {
+            Form {
+                ForEach(Array(settings.enumerated()), id: \.offset) { _, settings in
+                    let image = ruleBehaviorImages[settings.ruleBehavior]
+                    SnapSourceSettingsToggle(settings: settings, image: image)
+                }
+            }
+            .navigationTitle("Snap Sources")
+            .navigationBarTitleDisplayMode(.inline)
+            .toolbar {
+                ToolbarItem(placement: .confirmationAction) {
+                    Button("Done") { dismiss() }
+                }
+            }
+        }
+    }
+}
+
+/// A toggle for enabling and disabling a given snap source settings.
+private struct SnapSourceSettingsToggle: View {
+    /// The snap source settings to enable and disable.
+    let settings: SnapSourceSettings
+    
+    /// The image to use in the toggle's label.
+    let image: Image?
+    
+    /// A Boolean value indicating whether the toggle is enabled.
+    @State private var isEnabled = false
+    
+    var body: some View {
+        Toggle(isOn: $isEnabled) {
+            Label {
+                Text(settings.source.name)
+            } icon: {
+                image
+            }
+        }
+        .onChange(of: isEnabled) { newValue in
+            settings.isEnabled = newValue
+        }
+        .onAppear {
+            isEnabled = settings.isEnabled
+        }
+    }
+}
+
+// MARK: - Extensions
+
+extension SnapSource {
+    /// The name of the snap source.
+    var name: String {
+        switch self {
+        case let graphicsOverlay as GraphicsOverlay:
+            graphicsOverlay.id
+        case let layerContent as LayerContent:
+            layerContent.name
+        default:
+            "\(self)"
+        }
+    }
+}
+
+extension Optional<SnapRuleBehavior> {
+    fileprivate static var allCases: [Self] {
+        return [.none, .rulesLimitSnapping, .rulesPreventSnapping]
```

Comment:
interesting that we made "none" the nil case, but have to add none back in here for the UI purposes. Do you think we made the right choice?

Reply:
In samples we tend to show all the options including `nil` in the picker, so we need to add `Optional.none` back for the UI.

In real world apps, I'm seeing a trend of "toggle + picker" UI (e.g. Apple's Reminders app), where toggle-on shows the picker, and the picker shows all non-optional cases. In this sort of UI, toggle-off means `nil`, which backs up our decision in issue 3380 to make some enums `none` case -> optional.

Reply:
> Do you think we made the right choice?

Yes, absolutely. The none in this list is actually `Optional.none`. Being able to use it when dealing with an optional value is pretty cool. You can actually switch over an optional value and have a `.none` case as if `none` was a case of the enum itself, making the switch exhaustive.